internal/v5/archive_test.go

```diff
@@ -1729,3 +1756,125 @@ func (s *ArchiveSearchSuite) TestGetSearchUpdate(c *gc.C) {
+		Do:      bakeryDo(client),
+	})
+	c.Assert(rec.Code, gc.Equals, http.StatusOK)
+	c.Assert(rec.Body.Bytes(), gc.DeepEquals, archiveBytes)
+	c.Assert(rec.Header().Get(params.ContentHashHeader), gc.Equals, hashOfBytes(archiveBytes))
+	c.Assert(rec.Header().Get(params.EntityIdHeader), gc.Equals, "cs:~charmers/precise/terms-0")
+
+	// Check that the HTTP range logic is plugged in OK. If this
+	// is working, we assume that the whole thing is working OK,
+	// as net/http is well-tested.
+	rec = httptesting.DoRequest(c, httptesting.DoRequestParams{
+		Handler: s.srv,
+		URL:     archiveUrl,
+		Header:  http.Header{"Range": {"bytes=10-100"}},
+		Do:      bakeryDo(client),
+	})
+	c.Assert(rec.Code, gc.Equals, http.StatusPartialContent, gc.Commentf("body: %q", rec.Body.Bytes()))
+	c.Assert(rec.Body.Bytes(), gc.HasLen, 100-10+1)
+	c.Assert(rec.Body.Bytes(), gc.DeepEquals, archiveBytes[10:101])
+	c.Assert(rec.Header().Get(params.ContentHashHeader), gc.Equals, hashOfBytes(archiveBytes))
+	c.Assert(rec.Header().Get(params.EntityIdHeader), gc.Equals, "cs:~charmers/precise/terms-0")
+}
+
+func (s *ArchiveSuiteWithTerms) TestGetUserHasNotAgreedToTerms(c *gc.C) {
+	s.dischargeTerms = func(_, _ string) ([]checkers.Caveat, error) {
+		return nil, errgo.New("user has not agreed to specified terms and conditions")
+	}
+	client := httpbakery.NewHTTPClient()
+
+	id := newResolvedURL("cs:~charmers/precise/terms-0", -1)
+	s.assertUploadCharm(c, "POST", id, "terms")
+	err := s.store.SetPerms(&id.URL, "read", params.Everyone, id.URL.User)
+	c.Assert(err, gc.IsNil)
+
+	archiveUrl := storeURL("~charmers/precise/terms-0/archive")
+	httptesting.DoRequest(c, httptesting.DoRequestParams{
+		Handler:     s.srv,
+		URL:         archiveUrl,
+		Do:          bakeryDo(client),
+		ExpectError: ".*third party refused discharge: cannot discharge: user has not agreed to specified terms and conditions",
+	})
+}
+
+func (s *ArchiveSuiteWithTerms) TestGetIgnorningTermsWithBasicAuth(c *gc.C) {
+	s.dischargeTerms = func(_, _ string) ([]checkers.Caveat, error) {
+		return nil, errgo.New("user has not agreed to specified terms and conditions")
+	}
+
+	id := newResolvedURL("cs:~charmers/precise/terms-0", -1)
+	terms := s.assertUploadCharm(c, "POST", id, "terms")
+	err := s.store.SetPerms(&id.URL, "read", params.Everyone, id.URL.User)
+	c.Assert(err, gc.IsNil)
+
+	archiveBytes, err := ioutil.ReadFile(terms.Path)
+	c.Assert(err, gc.IsNil)
+
+	archiveUrl := storeURL("~charmers/precise/terms-0/archive")
+	rec := httptesting.DoRequest(c, httptesting.DoRequestParams{
+		Handler: s.srv,
+		URL:     archiveUrl,
+		Header:  basicAuthHeader(testUsername, testPassword),
+	})
+	c.Assert(rec.Code, gc.Equals, http.StatusOK)
+	c.Assert(rec.Body.Bytes(), gc.DeepEquals, archiveBytes)
+	c.Assert(rec.Header().Get(params.ContentHashHeader), gc.Equals, hashOfBytes(archiveBytes))
+	c.Assert(rec.Header().Get(params.EntityIdHeader), gc.Equals, "cs:~charmers/precise/terms-0")
+	assertCacheControl(c, rec.Header(), true)
```

Comment:
And again.
Perhaps consider making a assertArchiveDownload function that can be called by all these tests. Then it would be reasonable for the range checking to be in that, as it wouldn't be duplicated.